test/storage/bplustree_test.cpp

```diff
@@ -0,0 +1,1468 @@
+    int k = i % 1000;                  // there will be 100 inserts for same key
+    int v = rand_r(&globalseed) % 50;  // expect one duplicate value per key
+    if (keys_values.count(k) == 0) {
+      std::set<int> value_list;
+      value_list.insert(v);
+      keys_values[k] = value_list;
+    } else {
+      if (keys_values[k].count(v) == 0) {
+        keys_values[k].insert(v);
+      }
+    }
+    BPlusTree<int, int>::KeyElementPair p1;
+    p1.first = k;
+    p1.second = v;
+    bplustree->Insert(p1, predicate);
+  }
+  EXPECT_EQ(bplustree->DuplicateKeyValuesCheck(&keys_values), true);
+
+  delete bplustree;
+}
+void ScanKeyTest() {
+  auto predicate = [](const int slot) -> bool { return false; };
+
+  // Insert Keys
+  auto bplustree = new BPlusTree<int, int>;
+  bplustree->SetInnerNodeSizeUpperThreshold(5);
+  bplustree->SetLeafNodeSizeUpperThreshold(5);
+  std::unordered_map<int, std::set<int>> keys_values;
+  for (unsigned i = 0; i < 100000; i++) {
+    int k = rand_r(&globalseed) % 1000;
+    int v = rand_r(&globalseed) % 500000;
+    int is_value_unique = 1;
+    if (keys_values.count(k) == 0) {
+      std::set<int> value_list;
+      value_list.insert(v);
+      keys_values[k] = value_list;
+    } else {
+      if (keys_values[k].count(v) != 0) {
+        is_value_unique = 0;
+      } else {
+        keys_values[k].insert(v);
+      }
+    }
+    bool is_inserted = bplustree->Insert(BPlusTree<int, int>::KeyElementPair(k, v), predicate);
+    EXPECT_EQ(is_value_unique, is_inserted);
+  }
+  auto itr_map = keys_values.begin();
+  while (itr_map != keys_values.end()) {
+    int k = itr_map->first;
+    std::set<int> values = keys_values[k];
+    std::vector<int> result;
+    bplustree->FindValueOfKey(k, &result);
+    for (int i : result) {
+      EXPECT_EQ(values.count(i), 1);
+      values.erase(i);
+    }
+    EXPECT_EQ(values.size(), 0);
+    itr_map++;
+  }
+
+  delete bplustree;
+}
+
+void BasicBPlusTreeDeleteTestNoSplittingOfRoot() {
+  auto predicate = [](const TupleSlot slot) -> bool { return false; };
+
+  auto bplustree = new BPlusTree<int, TupleSlot>;
+  for (unsigned i = 0; i < 100; i++) {
+    BPlusTree<int, TupleSlot>::KeyElementPair p1;
+    p1.first = i;
+    bplustree->Insert(p1, predicate);
+  }
+
+  using ElementType = BPlusTree<int, TupleSlot>::KeyElementPair;
+
+  auto node = reinterpret_cast<BPlusTree<int, TupleSlot>::ElasticNode<ElementType> *>(bplustree->GetRoot());
+  unsigned i = 0;
+  for (ElementType *element_p = node->Begin(); element_p != node->End(); element_p++) {
+    EXPECT_EQ(element_p->first, i);
+    i++;
+  }
+  EXPECT_EQ(i, 100);
+
+  // Delete all values
+  for (i = 0; i < 100; i++) {
+    BPlusTree<int, TupleSlot>::KeyElementPair p1;
+    p1.first = i;
+    bplustree->DeleteElement(p1);
+    EXPECT_EQ(bplustree->IsPresent(i), false);
+  }
+  EXPECT_EQ(bplustree->GetRoot() == nullptr, true);
+
+  delete bplustree;
+}
+
+void LargeKeySequentialInsertAndRetrievalTest() {
```

Comment:
Can we comment on why this is considered a "Larger" test? The numbers do seem large, but I think it would be more clear to clarify this.

Reply:
Done